### ISSUE TITLE
Buffer settings for a very slow connection

### ISSUE THREAD

**Issue by SieldP** [cedrozor/myrtille#113]
Hello,
First of all, I would like to apologize for my english which is not perfect and I would like to say that  I am very glad to use Myrtille, it's an awesome and very useful project ! :)

My problem is that I have a slow internet connection (about 0.5M in upload and 2 in download) Myrtille latency : from 70ms where buffer is empty, to 800ms when the buffer is overloaded)

When I use Myrtille (configured with the WebSocket), I have some problem displaying images.
For example: when I'm on a video or scrolling a web page, there are too many images that fill the buffer. And as long as these images are not displayed, the last images will not appear.
Even if I click on reduce the window: the window has been reduced instantly on the server, but I have to wait for the images of the buffer to appear to see the window reduced, or I must disconnect from the session rdp and start another one.

I know I can set the buffer in the config.js file. I would like to set it for that my orders (window reduction for example) appear instantly on the internet browser.
In your opinion, what are the best parameters (bufferDelayBase, bufferDelayEmpty and bufferSize) in my case ? Ss that the buffer is constantly emptied and the last update of images are displayed quickly.
Or can you explain more clearly what is the bufferDelayBase and bufferDelayEmpty parameters so that I can determine the best parameters?
I would really like my commands to be instant, even if I have to reduce frames to 5 FPS.

Thank you very much :)

**Comment by cedrozor** [comment]
Hi,

It's more a display updates issue than user inputs. Into config.js, default settings are:

```
var displayMode = displayModeEnum.AUTO;
var imageEncoding = imageEncodingEnum.PNG;
var imageQuality = 100;
var imageQuantity = 100;
```
You may want to change them to:

```
var displayMode = displayModeEnum.AUTO;
var imageEncoding = imageEncodingEnum.**AUTO**;
var imageQuality = 25;
var imageQuantity = 50;
```
It will optimize the bandwidth usage (encode to JPEG or PNG dynamically, depending on the content), cap the JPEG quality to 25% (the display will be less crisp but good enough most of the time) and halve your framerate, but myrtille will be more responsive overall.

**Comment by SieldP** [comment]
Thank you, It looks a bit better now..

imageEncodingEnum.AUTO parameter choses automaticly the image file formats depending of its size ?
However, I noticed that it never switch to WEBP format, although it's usually the smalest format.

Last but not least can the buffer reactivity of myrtille ?

**Comment by cedrozor** [comment]
Yes, AUTO encodes in both JPEG and PNG and send the lowest sized image. It doesn't encode in WEBP because having 3 passes for each image starts to kill the responsiveness and also increase the cpu usage.

You can specifically choose `var imageEncoding = imageEncodingEnum.WEBP;` if you feel it's best for your usage.

The inputs buffer duration auto-fits to latency (the more latency = the more buffering, and inversely).

If you want to increase the buffer size, replace into config.js:
`var bufferSize = 128;`
by
`var bufferSize = 256;` (double buffer size)

But low upload bandwidth isn't usually an issue for inputs; average inputs bandwidth is below 0,5KB/s and you will have difficulties going more than 2 or 3KB/s. Each input is about 10 bytes, so a full 256 items buffer is about 256 * 10 = 2560 bytes ~= 2,5KB

You should enable the connection info bar (hidden by default) by commenting #stat into default.css : 

```
/* #stat, */
#debug,
#browser
{
    visibility: hidden;
    display: none;
}
```

Then click the "Stat" button into the toolbar. It will show useful info about latency, buffering, bandwidth and more :)

**Comment by SieldP** [comment]
Thank a lot,

I just uderstood that the buffer is an input buffering, and not display my bad ... :p
However, I think that is a good idea to add an display buffering: using WebSocket, you can have a region updates buffer ... -> When an new region update is available and covering the position of an old region update still on the buffer, you can remove the old one from the buffer to improve the latency. This will avoid a possible rolling-back of the display :) I think that's what mstsc.exe uses.

Thank again

**Comment by cedrozor** [comment]
Yes, for sure, I thought about buffering updates in the past (we even had a clipping algorithm for our PoC with rdesktop) but it's already done by wfreerdp, the same way mstsc does (and even windows itself) so I didn't felt the need for a double bufferization (both into wfreerdp and myrtille).

There is a small cache into myrtille, but it's used with HTML4 browsers when polling updates. Websockets are fast enough to push updates on-the-fly.

Thanks for your insight! that's said, I'm sure there are still room for optimizations and I would be glad to hear about them :)

**Comment by SieldP** [comment]
The problem with FreeRDP's buffering, is that the freerdp latency is non-existent, because using FreeRDP with Myrtille, the FreeRDP connection is maked locally.
However, I think that the buffer should be in Myrtille because it's who sends information through internet ...
I compared both, mstsc.exe and Myrtille through a slow internet connection, and I feel that the myrtille buffer is missing : mstsc.exe better fit the latency

**Comment by cedrozor** [comment]
It's true the most latency is between the browser and the gateway. I tried many ways to handle it, but wasn't satisfied by any of them:
- when using websockets, have an acknowledgment mechanism for a display update before sending the next one; so that, if the connection or browser is struggling under the load, it helps to mitigate it
- have a flag indicating a latency issue (there is already such thing client side), send it to the gateway and stop sending updates as long as this flag is up
- polling display updates from a cache (buffer), instead of pushing them

The 3rd solution is used in xhr mode (html4); perhaps a good solution for websockets would be to have a buffer with a limited duration (and size) and unstack it according to latency.

I'll try that! thanks for your help :)

**Comment by SieldP** [comment]
You're welcome :)

I'm sure that it will improve performance :)

thanks you again

**Comment by cedrozor** [comment]
H Sield,

I implemented a websocket buffer on the gateway, which also relies on a time limited cache for an even better performance :) it's available into myrtille 2.0.0 (I just released it).

Could you re-check and tell me if you see an improvement ?
